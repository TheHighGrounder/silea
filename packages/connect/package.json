{
	"name": "@sileajs/connect",
	"version": "0.0.2",
	"description": "The state manager of Silea. Similar to MobX but much more lightweight because it's based on ES6 proxies.",
	"keywords": [
		"silea",
		"connect",
		"state-manager"
	],
	"homepage": "https://sileajs.com",
	"license": "SEE LICENSE IN LICENSE",
	"main": "src/index.js",
	"types": "src/index.d.ts",
	"repository": {
		"type": "git",
		"url": "https://github.com/SileaJS/silea.git"
	},
	"scripts": {
		"test:ci": "../../node_modules/.bin/jest --ci --coverage",
		"test": "../../node_modules/.bin/jest --watch",
		"test:inspect": "node --inspect ../../node_modules/.bin/jest --watch --no-cache --runInBand"
	},
	"dependencies": {
		"@sileajs/error": "0.0.2"
	},
	"devDependencies": {
		"react-test-renderer": "16.13.1"
	},
	"publishConfig": {
		"access": "public"
	}
}
