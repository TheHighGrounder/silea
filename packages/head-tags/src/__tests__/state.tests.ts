/* eslint-disable @typescript-eslint/camelcase */
import clone from "clone-deep";
import { createStore, InitializedStore } from "@sileajs/connect";
import wpSource from "@sileajs/wp-source/src";
import HeadTagsPackage, { State, HeadTags } from "../../types";
import headTagsPackage from "..";
import {
	mockPostEntity,
	mockPostType,
	mockTaxonomy,
	mockAuthor,
} from "./mocks/utils";

// Spy `console.warn`
const warn = jest.spyOn(global.console, "warn");

// Mock Silea state.
let store: InitializedStore<HeadTagsPackage>;
beforeEach(() => {
	warn.mockClear();

	// Create store.
	const config: HeadTagsPackage = clone(headTagsPackage());

	// Mock wp-source state.
	config.state.source = clone(wpSource()).state.source;
	config.state.source.api = "https://test.sileajs.com/wp-json";
	// Mock router state.
	config.state.router = { link: "/", state: {} };
	// Mock site url.
	config.state.silea = { url: "https://mars.sileajs.com" };
	// Initialize store.
	store = createStore(config);
});

describe("state.headTags.get() (post entity)", () => {
	const setUpState = (state: State, headTags?: HeadTags) => {
		// Populate source state.
		const { post, data } = mockPostEntity(headTags);
		state.source.post = post;
		state.source.data = data;

		// Populate router state.
		state.router.link = "/post-1/";
	};

	test("returns an empty array if `head_tags` is undefined", () => {
		const headTags = undefined;
		// Populate all state.
		setUpState(store.state, headTags);
		// Test current head tags.
		expect(store.state.headTags.get(store.state.router.link)).toEqual([]);
	});

	test("returns elements without attributes (title)", () => {
		const headTags: HeadTags = [
			{ tag: "title", content: "Post 1 - Silea" },
		];
		// Populate all state.
		setUpState(store.state, headTags);
		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("returns elements without content (meta tags)", () => {
		const headTags: HeadTags = [
			{
				tag: "meta",
				attributes: {
					name: "robots",
					content:
						"max-snippet:-1, max-image-preview:large, max-video-preview:-1",
				},
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);
		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("transforms links that point to WordPress pages", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://test.sileajs.com/post-1/",
				},
			},
			{
				tag: "meta",
				attributes: {
					property: "og:url",
					content: "https://test.sileajs.com/post-1/",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "shortlink",
					href: "https://test.sileajs.com/?p=1",
				},
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);
		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("doesn't change links that don't point to WordPress pages", () => {
		const headTags: HeadTags = [
			{
				tag: "meta",
				attributes: {
					property: "og:image",
					content:
						"https://test.sileajs.com/wp-content/uploads/2019/12/img.jpg",
				},
			},
			{
				tag: "meta",
				attributes: {
					name: "twitter:image",
					content:
						"https://test.sileajs.com/wp-content/uploads/2019/12/img.jpg",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "dns-prefetch",
					href: "//s.w.org",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "alternate",
					type: "application/rss+xml",
					title: "silea » Feed",
					href: "https://test.sileajs.com/feed/",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "alternate",
					type: "application/rss+xml",
					title: "silea » Comments Feed",
					href: "https://test.sileajs.com/comments/feed/",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "alternate",
					type: "application/rss+xml",
					title: "silea » Post 1 Comments Feed",
					href: "https://test.sileajs.com/post-1/feed/",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "https://api.w.org/",
					href: "https://test.sileajs.com/wp-json/",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "EditURI",
					type: "application/rsd+xml",
					title: "RSD",
					href: "https://test.sileajs.com/xmlrpc.php?rsd",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "wlwmanifest",
					type: "application/wlwmanifest+xml",
					href:
						"https://test.sileajs.com/wp-includes/wlwmanifest.xml",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "alternate",
					type: "application/json+oembed",
					href:
						"https://test.sileajs.com/wp-json/oembed/1.0/embed?url=http%3A%2F%2Ftest.sileajs.com%2Fpost-1%2F",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "alternate",
					type: "text/xml+oembed",
					href:
						"https://test.sileajs.com/wp-json/oembed/1.0/embed?url=http%3A%2F%2Ftest.sileajs.com%2Fpost-1%2F&format=xml",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "pingback",
					href: "https://test.sileajs.com/xmlrpc.php",
				},
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);
		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("transform links inside ld+json data", () => {
		const headTags: HeadTags = [
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
					class: "yoast-schema-graph yoast-schema-graph--main",
				},
				content: JSON.stringify({
					"@context": "https://schema.org",
					"@graph": [
						{
							// All these links should change...
							"@type": "WebSite",
							"@id": "https://test.sileajs.com/#website",
							url: "https://test.sileajs.com/",
							potentialAction: {
								"@type": "SearchAction",
								target:
									"https://test.sileajs.com/?s={search_term_string}",
							},
						},
						{
							"@type": "ImageObject",
							"@id":
								"https://test.sileajs.com/post-1/#primaryimage",
							// ...except this one.
							url:
								"https://test.sileajs.com/wp-content/uploads/2019/12/img.jpg",
						},
					],
				}),
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);
		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("transforms links appropiately when WP is in a subdirectory", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://test.sileajs.com/subdir/post-1/", // should change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "https://api.w.org/",
					href: "https://test.sileajs.com/subdir/wp-json/", // should not change
				},
			},
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
				},
				content: JSON.stringify({
					"@context": "https://schema.org",
					"@graph": [
						{
							// All these links should change
							"@type": "WebSite",
							"@id":
								"https://test.sileajs.com/subdir/#website",
							url: "https://test.sileajs.com/subdir/",
						},
					],
				}),
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);

		// Change API prop
		store.state.source.api = "https://test.sileajs.com/subdir/wp-json/";

		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("transforms links appropiately when Silea is in a subdirectory", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://test.sileajs.com/subdir/post-1/", // should change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "https://api.w.org/",
					href: "https://test.sileajs.com/subdir/wp-json/", // should not change
				},
			},
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
				},
				content: JSON.stringify({
					"@context": "https://schema.org",
					"@graph": [
						{
							// All these links should change
							"@type": "WebSite",
							"@id":
								"https://test.sileajs.com/subdir/#website",
							url: "https://test.sileajs.com/subdir/",
						},
					],
				}),
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);

		// Change API and Silea URL prop
		store.state.silea.url = "https://sileajs.com/mars/";
		store.state.source.api = "https://test.sileajs.com/subdir/wp-json/";

		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("shows a warning message if `state.silea.url` is not defined", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://test.sileajs.com/subdir/post-1/", // should change
				},
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);

		// Remove `state.silea.url`.
		delete store.state.silea.url;

		const warn = jest.spyOn(global.console, "warn");

		// Test current head tags.
		store.state.headTags.get(store.state.router.link);

		expect(warn).toHaveBeenCalled();
		expect(warn.mock.calls).toMatchSnapshot();
	});

	test('shows a warning message if a <script type="ld+json"> could not be parsed', () => {
		const headTags: HeadTags = [
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
				},
				content: `{
  "@context": "https://schema.org",
  "@graph": [
    {
      "@type": "WebPage",
      "@id": "https://test.sileajs.com/wrong-title/#webpage",
      "url": "https://test.sileajs.com/wrong-title/",
      "inLanguage": "en-US",
      "name": "Wrong title with "quotes" - silea",
      "isPartOf": {
        "@id": "https://test.sileajs.com/#website"
      }
    }
  ]
}`,
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);

		// Test current head tags.
		store.state.headTags.get(store.state.router.link);

		expect(warn).toHaveBeenCalled();
		expect(warn.mock.calls).toMatchSnapshot();
	});

	test("doesn't transform links if `state.headTags.transformLinks` = false", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://test.sileajs.com/post-1/", // should not change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "https://api.w.org/",
					href: "https://test.sileajs.com/wp-json/", // should not change
				},
			},
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
				},
				content: JSON.stringify({
					"@context": "https://schema.org",
					"@graph": [
						{
							// All these links should not change
							"@type": "WebSite",
							"@id": "https://test.sileajs.com/#website",
							url: "https://test.sileajs.com/",
						},
					],
				}),
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);

		// Set `transformLinks` to false.
		store.state.headTags.transformLinks = false;

		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("doesn't transform links if `source.api` and `silea.url` are the same site", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://test.sileajs.com/post-1/", // should not change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "https://api.w.org/",
					href: "https://test.sileajs.com/wp-json/", // should not change
				},
			},
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
				},
				content: JSON.stringify({
					"@context": "https://schema.org",
					"@graph": [
						{
							// All these links should not change
							"@type": "WebSite",
							"@id": "https://test.sileajs.com/#website",
							url: "https://test.sileajs.com/",
						},
					],
				}),
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);

		// Set `silea.url` to the same site.
		store.state.silea.url = "https://test.sileajs.com/";

		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("transform links using a custom `base` value", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://different.sileajs.com/blog/post-1/", // should change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "https://api.w.org/",
					href: "https://test.sileajs.com/wp-json/", // should not change
				},
			},
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
				},
				content: JSON.stringify({
					"@context": "https://schema.org",
					"@graph": [
						{
							// All these links should change
							"@type": "WebSite",
							"@id":
								"https://different.sileajs.com/blog/#website",
							url: "https://different.sileajs.com/blog/",
						},
					],
				}),
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);

		// Set `transformLinks` to false.
		const { transformLinks } = store.state.headTags;
		if (transformLinks)
			transformLinks.base = "https://different.sileajs.com/blog/";

		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});

	test("transform links using a custom `ignore` value", () => {
		const headTags: HeadTags = [
			{
				tag: "meta",
				attributes: {
					name: "twitter:image",
					// should change
					content:
						"https://test.sileajs.com/wp-content/uploads/2019/12/img.jpg",
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "alternate",
					type: "application/rss+xml",
					title: "silea » Feed",
					// SHOULD NOT CHANGE
					href: "https://test.sileajs.com/do-not-change/me/",
				},
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);

		// Set `transformLinks` to false.
		const { transformLinks } = store.state.headTags;
		if (transformLinks) {
			transformLinks.base = "https://test.sileajs.com/";
			transformLinks.ignore = "do\\-not\\-change";
		}

		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});
});

describe("state.headTags.get() (post type)", () => {
	const setUpState = (state: State, headTags?: HeadTags) => {
		// Populate source state.
		const { type, data } = mockPostType(headTags);
		state.source.type = type;
		state.source.data = data;

		// Populate router state.
		state.router.link = "/";
	};

	test("works with post types", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://test.sileajs.com/", // should change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "next",
					href: "https://test.sileajs.com/page/2/", // should change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "https://api.w.org/",
					href: "https://test.sileajs.com/wp-json/", // should not change
				},
			},
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
				},
				content: JSON.stringify({
					"@context": "https://schema.org",
					"@graph": [
						{
							// All these links should change
							"@type": "WebSite",
							"@id": "https://test.sileajs.com/#website",
							url: "https://test.sileajs.com/",
						},
					],
				}),
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);
		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});
});

describe("state.headTags.get() (taxonomy)", () => {
	const setUpState = (state: State, headTags?: HeadTags) => {
		// Populate source state.
		const { category, data } = mockTaxonomy(headTags);
		state.source.category = category;
		state.source.data = data;

		// Populate router state.
		state.router.link = "/category/cat-1/";
	};

	test("works with taxonomies", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://test.sileajs.com/category/cat-1/", // should change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "next",
					href: "https://test.sileajs.com/category/cat-1/page/2/", // should change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "https://api.w.org/",
					href: "https://test.sileajs.com/wp-json/", // should not change
				},
			},
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
				},
				content: JSON.stringify({
					"@context": "https://schema.org",
					"@graph": [
						{
							// All these links should change
							"@type": "WebSite",
							"@id": "https://test.sileajs.com/#website",
							url: "https://test.sileajs.com/",
						},
					],
				}),
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);
		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});
});

describe("state.headTags.get() (author)", () => {
	const setUpState = (state: State, headTags?: HeadTags) => {
		// Populate source state.
		const { author, data } = mockAuthor(headTags);
		state.source.author = author;
		state.source.data = data;

		// Populate router state.
		state.router.link = "/author/author-1/";
	};

	test("works with auhors", () => {
		const headTags: HeadTags = [
			{
				tag: "link",
				attributes: {
					rel: "canonical",
					href: "https://test.sileajs.com/author/author-1/", // should change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "next",
					href:
						"https://test.sileajs.com/author/author-1/page/2/", // should change
				},
			},
			{
				tag: "link",
				attributes: {
					rel: "https://api.w.org/",
					href: "https://test.sileajs.com/wp-json/", // should not change
				},
			},
			{
				tag: "script",
				attributes: {
					type: "application/ld+json",
				},
				content: JSON.stringify({
					"@context": "https://schema.org",
					"@graph": [
						{
							// All these links should change
							"@type": "WebSite",
							"@id": "https://test.sileajs.com/#website",
							url: "https://test.sileajs.com/",
						},
					],
				}),
			},
		];
		// Populate all state.
		setUpState(store.state, headTags);
		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});
});

describe("state.headTags.get() (no entity)", () => {
	const setUpState = (state: State) => {
		// Populate source state.
		state.source.data = {
			"/2019/12/": {
				items: [],
				isArchive: true,
				isDate: true,
				year: 2019,
				month: 12,
				day: undefined,
				isFetching: false,
				isReady: true,
			},
		};

		// Populate router state.
		state.router.link = "/2019/12/";
	};

	test("returns an empty array if there is no entity for current link", () => {
		// Populate all state.
		setUpState(store.state);
		// Test current head tags.
		expect(
			store.state.headTags.get(store.state.router.link)
		).toMatchSnapshot();
	});
});
