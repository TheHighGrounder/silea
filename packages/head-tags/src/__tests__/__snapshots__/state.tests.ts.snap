// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state.headTags.get() (author) works with auhors 1`] = `
Array [
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/author/author-1/",
      "rel": "canonical",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/author/author-1/page/2/",
      "rel": "next",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-json/",
      "rel": "https://api.w.org/",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "type": "application/ld+json",
    },
    "content": "{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"@type\\":\\"WebSite\\",\\"@id\\":\\"https://mars.sileajs.com/#website\\",\\"url\\":\\"https://mars.sileajs.com/\\"}]}",
    "tag": "script",
  },
]
`;

exports[`state.headTags.get() (no entity) returns an empty array if there is no entity for current link 1`] = `Array []`;

exports[`state.headTags.get() (post entity) doesn't change links that don't point to WordPress pages 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": "https://test.sileajs.com/wp-content/uploads/2019/12/img.jpg",
      "property": "og:image",
    },
    "tag": "meta",
  },
  Object {
    "attributes": Object {
      "content": "https://test.sileajs.com/wp-content/uploads/2019/12/img.jpg",
      "name": "twitter:image",
    },
    "tag": "meta",
  },
  Object {
    "attributes": Object {
      "href": "//s.w.org",
      "rel": "dns-prefetch",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/feed/",
      "rel": "alternate",
      "title": "silea » Feed",
      "type": "application/rss+xml",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/comments/feed/",
      "rel": "alternate",
      "title": "silea » Comments Feed",
      "type": "application/rss+xml",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/post-1/feed/",
      "rel": "alternate",
      "title": "silea » Post 1 Comments Feed",
      "type": "application/rss+xml",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-json/",
      "rel": "https://api.w.org/",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/xmlrpc.php?rsd",
      "rel": "EditURI",
      "title": "RSD",
      "type": "application/rsd+xml",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-includes/wlwmanifest.xml",
      "rel": "wlwmanifest",
      "type": "application/wlwmanifest+xml",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-json/oembed/1.0/embed?url=http%3A%2F%2Ftest.sileajs.com%2Fpost-1%2F",
      "rel": "alternate",
      "type": "application/json+oembed",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-json/oembed/1.0/embed?url=http%3A%2F%2Ftest.sileajs.com%2Fpost-1%2F&format=xml",
      "rel": "alternate",
      "type": "text/xml+oembed",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/xmlrpc.php",
      "rel": "pingback",
    },
    "tag": "link",
  },
]
`;

exports[`state.headTags.get() (post entity) doesn't transform links if \`source.api\` and \`silea.url\` are the same site 1`] = `
Array [
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/post-1/",
      "rel": "canonical",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-json/",
      "rel": "https://api.w.org/",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "type": "application/ld+json",
    },
    "content": "{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"@type\\":\\"WebSite\\",\\"@id\\":\\"https://test.sileajs.com/#website\\",\\"url\\":\\"https://test.sileajs.com/\\"}]}",
    "tag": "script",
  },
]
`;

exports[`state.headTags.get() (post entity) doesn't transform links if \`state.headTags.transformLinks\` = false 1`] = `
Array [
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/post-1/",
      "rel": "canonical",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-json/",
      "rel": "https://api.w.org/",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "type": "application/ld+json",
    },
    "content": "{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"@type\\":\\"WebSite\\",\\"@id\\":\\"https://test.sileajs.com/#website\\",\\"url\\":\\"https://test.sileajs.com/\\"}]}",
    "tag": "script",
  },
]
`;

exports[`state.headTags.get() (post entity) returns elements without attributes (title) 1`] = `
Array [
  Object {
    "content": "Post 1 - Silea",
    "tag": "title",
  },
]
`;

exports[`state.headTags.get() (post entity) returns elements without content (meta tags) 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": "max-snippet:-1, max-image-preview:large, max-video-preview:-1",
      "name": "robots",
    },
    "tag": "meta",
  },
]
`;

exports[`state.headTags.get() (post entity) shows a warning message if \`state.silea.url\` is not defined 1`] = `
Array [
  Array [
    "Property \`state.headTags.links.transform\` is defined but \`state.silea.url\` is not. All links in <head> tags pointing to other site (e.g. WordPress) instead to the Silea site won't be changed.

If you have any questions, join our community at https://community.sileajs.com/.
",
  ],
]
`;

exports[`state.headTags.get() (post entity) shows a warning message if a <script type="ld+json"> could not be parsed 1`] = `
Array [
  Array [
    "The following content of a <script type=\\"ld+json\\"> tag is not a valid JSON. Links in that tag will not be changed.

{
  \\"@context\\": \\"https://schema.org\\",
  \\"@graph\\": [
    {
      \\"@type\\": \\"WebPage\\",
      \\"@id\\": \\"https://test.sileajs.com/wrong-title/#webpage\\",
      \\"url\\": \\"https://test.sileajs.com/wrong-title/\\",
      \\"inLanguage\\": \\"en-US\\",
      \\"name\\": \\"Wrong title with \\"quotes\\" - silea\\",
      \\"isPartOf\\": {
        \\"@id\\": \\"https://test.sileajs.com/#website\\"
      }
    }
  ]
}

If you have any questions, join our community at https://community.sileajs.com/.
",
  ],
]
`;

exports[`state.headTags.get() (post entity) transform links inside ld+json data 1`] = `
Array [
  Object {
    "attributes": Object {
      "class": "yoast-schema-graph yoast-schema-graph--main",
      "type": "application/ld+json",
    },
    "content": "{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"@type\\":\\"WebSite\\",\\"@id\\":\\"https://mars.sileajs.com/#website\\",\\"url\\":\\"https://mars.sileajs.com/\\",\\"potentialAction\\":{\\"@type\\":\\"SearchAction\\",\\"target\\":\\"https://mars.sileajs.com/?s={search_term_string}\\"}},{\\"@type\\":\\"ImageObject\\",\\"@id\\":\\"https://mars.sileajs.com/post-1/#primaryimage\\",\\"url\\":\\"https://test.sileajs.com/wp-content/uploads/2019/12/img.jpg\\"}]}",
    "tag": "script",
  },
]
`;

exports[`state.headTags.get() (post entity) transform links using a custom \`base\` value 1`] = `
Array [
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/post-1/",
      "rel": "canonical",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-json/",
      "rel": "https://api.w.org/",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "type": "application/ld+json",
    },
    "content": "{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"@type\\":\\"WebSite\\",\\"@id\\":\\"https://mars.sileajs.com/#website\\",\\"url\\":\\"https://mars.sileajs.com/\\"}]}",
    "tag": "script",
  },
]
`;

exports[`state.headTags.get() (post entity) transform links using a custom \`ignore\` value 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": "https://mars.sileajs.com/wp-content/uploads/2019/12/img.jpg",
      "name": "twitter:image",
    },
    "tag": "meta",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/do-not-change/me/",
      "rel": "alternate",
      "title": "silea » Feed",
      "type": "application/rss+xml",
    },
    "tag": "link",
  },
]
`;

exports[`state.headTags.get() (post entity) transforms links appropiately when Silea is in a subdirectory 1`] = `
Array [
  Object {
    "attributes": Object {
      "href": "https://sileajs.com/mars/post-1/",
      "rel": "canonical",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/subdir/wp-json/",
      "rel": "https://api.w.org/",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "type": "application/ld+json",
    },
    "content": "{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"@type\\":\\"WebSite\\",\\"@id\\":\\"https://sileajs.com/mars/#website\\",\\"url\\":\\"https://sileajs.com/mars/\\"}]}",
    "tag": "script",
  },
]
`;

exports[`state.headTags.get() (post entity) transforms links appropiately when WP is in a subdirectory 1`] = `
Array [
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/post-1/",
      "rel": "canonical",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/subdir/wp-json/",
      "rel": "https://api.w.org/",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "type": "application/ld+json",
    },
    "content": "{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"@type\\":\\"WebSite\\",\\"@id\\":\\"https://mars.sileajs.com/#website\\",\\"url\\":\\"https://mars.sileajs.com/\\"}]}",
    "tag": "script",
  },
]
`;

exports[`state.headTags.get() (post entity) transforms links that point to WordPress pages 1`] = `
Array [
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/post-1/",
      "rel": "canonical",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "content": "https://mars.sileajs.com/post-1/",
      "property": "og:url",
    },
    "tag": "meta",
  },
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/?p=1",
      "rel": "shortlink",
    },
    "tag": "link",
  },
]
`;

exports[`state.headTags.get() (post type) works with post types 1`] = `
Array [
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/",
      "rel": "canonical",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/page/2/",
      "rel": "next",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-json/",
      "rel": "https://api.w.org/",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "type": "application/ld+json",
    },
    "content": "{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"@type\\":\\"WebSite\\",\\"@id\\":\\"https://mars.sileajs.com/#website\\",\\"url\\":\\"https://mars.sileajs.com/\\"}]}",
    "tag": "script",
  },
]
`;

exports[`state.headTags.get() (taxonomy) works with taxonomies 1`] = `
Array [
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/category/cat-1/",
      "rel": "canonical",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://mars.sileajs.com/category/cat-1/page/2/",
      "rel": "next",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "href": "https://test.sileajs.com/wp-json/",
      "rel": "https://api.w.org/",
    },
    "tag": "link",
  },
  Object {
    "attributes": Object {
      "type": "application/ld+json",
    },
    "content": "{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"@type\\":\\"WebSite\\",\\"@id\\":\\"https://mars.sileajs.com/#website\\",\\"url\\":\\"https://mars.sileajs.com/\\"}]}",
    "tag": "script",
  },
]
`;
